### PR TITLE
(need help) plugins/nvim-spectre: init + test 

### PR DESCRIPTION
**Pull request by hmajid2301**
Adds [nvim-spectre](https://github.com/nvim-pack/nvim-spectre).

I got about halfway and now am stuck with typing, any help would be greatly appreciated thanks.

### CODE REVIEW

File: plugins/utils/spectre.nix [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+      lineSep = helpers.defaultNullOpts.mkStr "└-----------------------------------------" "Shown at the bottom of the search results";
+
+      highlight = {
+        ui = helpers.defaultNullOpts.mkStr "String";
+        search = helpers.defaultNullOpts.mkStr "DiffChange";
+        replace = helpers.defaultNullOpts.mkStr "DiffDelete";
+      };
+
+      mappings =
+        option
```
Comment (hmajid2301) [inline plugins/utils/spectre.nix:41]: Im not really sure what to type this tbh, i created a type above but am not confident its close to being correct.

**Comment by hmajid2301** [comment]
Full config here: https://github.com/nvim-pack/nvim-spectre#customization

File: plugins/utils/spectre.nix [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+          cmd = helpers.defaultNullOpts.mkStr "sed" "Must be one of the options in the replaceEngine";
+          options = helpers.defaultNullOpts.mkNullable (with types; listOf str) "[]";
+        };
+      };
+      replaceVimCmd = helpers.defaultNullOpts.mkStr "cdo";
+      isOpenTargetWin = helpers.defaultNullOpts.mkBool true "Open file on opener window";
+      isInsertMode = helpers.defaultNullOpts.mkBool false "If set to true will open window in insert mode";
+      isBlockUIBreak = helpers.defaultNullOpts.mkBool false "If set to true will map backspace and enter key to avoid ui break";
+      openTemplate =
+        mkMappingsOption
```
Comment (hmajid2301) [inline plugins/utils/spectre.nix:224]: Also not sure what to actually use as a type here, any help would be appreciated

File: plugins/utils/spectre.nix [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    };
+
+  config = let
+    cfg = config.plugins.spectre;
+  in
+    mkIf cfg.enable {
+      extraPlugins = [cfg.package];
+
+      extraConfigLua = let
+        opts = with cfg; {
```
Comment (hmajid2301) [inline plugins/utils/spectre.nix:240]: will add options once the above is done

File: plugins/utils/spectre.nix [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            '';
+        };
+      };
+      default = {
+        find = {
+          cmd = helpers.defaultNullOpts.mkStr "rg" "Must be one of the options in the findEngine";
+          options = helpers.defaultNullOpts.mkNullable (with types; listOf str) "[ignore-case]";
+        };
+        replace = {
+          cmd = helpers.defaultNullOpts.mkStr "sed" "Must be one of the options in the replaceEngine";
```
Comment (hmajid2301) [inline plugins/utils/spectre.nix:215]: is there way to make it so it can only be an option from replaceEngine

**Comment by GaetanLepage** [comment]
Sorry for taking this long. I created the module from scratch in #1447  before finding this PR...

**Comment by hmajid2301** [comment]
no worries, you are doing this in your free time :).  It'll good to see yours and compare where I got stuck!

**Pull request closed without merging**